### PR TITLE
generic fetcher: Prefetch files

### PR DESCRIPTION
**Pull request by kosciCZ**
This PR implements the main parts of the generic fetcher, including checksum verification and integration test suite verifying the functionality. The fetcher does not yet produce SBOM components, which is intended and will be implemented in a follow-up PR. I can also extend this PR with the SBOM part if preferred, please let me know. 

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py:1]: Fixed.

File: cachi2/core/package_managers/generic/main.py
Comment (slimreaper35) [inline cachi2/core/package_managers/generic/main.py:84]: This does not make sense to me. There is an attribute in the class that is Optional and is not set when creating the object. It is only overwritten here. In addition, you try to do a comparison via `has_same_resolved_path` even if it is None.

File: cachi2/core/package_managers/generic/models.py
Comment (slimreaper35) [inline cachi2/core/package_managers/generic/models.py:20]: nitpick: I think it is worth adding a short description for all attributes in the class.

File: tests/integration/test_generic.py
Comment (kosciCZ) [reply tests/integration/test_generic.py:27]: Oh, cool. Will fix!

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py:61]: I think that was intended. From what I observed, `output_dir` is the cachi2 output folder. But the generic dependencies should go under `{output_dir}/DEFAULT_DEPS_DIR` and nowhere else, so it made sense to me to re_root. If I just used `join_within_root`, I could still set targets outside of the deps dir, which I don't think should be permitted.

File: cachi2/core/package_managers/generic/models.py
Comment (brunoapimentel) [inline cachi2/core/package_managers/generic/models.py:53]: I'm getting an error when trying to run Cachi2 on a sample repo I have locally, I believe this line is the culprit:
```
raise PydanticSchemaGenerationError(
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for os.PathLike[str].
 Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.
```

File: cachi2/core/package_managers/generic/main.py
Comment (brunoapimentel) [reply cachi2/core/package_managers/generic/main.py:84]: I feel like the `GenericArtifact` class is not bringing too much value, it seems mostly redundant. Maybe we can add the `resolved_path` attribute in the `LockfileArtifact` class itself, or even rely on `pathlib.Path(GenericArtifact.target).resolve()` to look for conflicts and drop this attribute altogether.

File: cachi2/core/package_managers/generic/main.py
Comment (brunoapimentel) [inline cachi2/core/package_managers/generic/main.py:76]: I have a small suggestion to avoid needing this nested loop:
```
existing_paths = set()

for artifact in artifacts:
   destination = output_dir.join_within_root(artifact.resolved_target)
   
   if destination in existing_paths:
      raise Exception()
   else:
       existing_paths.add(destination)     
```

And ideally keep the result of `destination` to avoid having to recalculate it during the download phase. I think this check could fit well into a tiny utility function.

File: tests/integration/test_generic.py
Comment (kosciCZ) [reply tests/integration/test_generic.py]: Added and moved into a separate test function that also builds the image.

File: tests/integration/test_generic.py
Comment (kosciCZ) [reply tests/integration/test_generic.py]: Fixed.

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: I used the general idea of this approach. Destination is now `_resolved_path` in the `LockfileArtifact`

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py]: Removed this class.

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py]: Added

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Removed the `GenericArtifact` class, it was mostly redundant. I originally created it, because I did not want to mix validation and whatever happens next to the artifact. I've added  `_resolved_path` in the `LockfileArtifact` class as a private attribute, so it couldn't be loaded from the lockfile, but could still be used to store the path. Thanks for the feedback, I've definitely overengineered it.

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py:60]: Resolving. If you still disagree, or have further questions, feel free to unresolve.

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py]: Removed entirely with the whole class.

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py]: Thanks for the tip! Changed.

File: cachi2/core/package_managers/generic/main.py
Comment (eskultety) [inline cachi2/core/package_managers/generic/main.py:116]: nitpick: This should be a separate commit as it's a unrelated change fixing a wrong type of error used for the given scenario.

File: cachi2/core/package_managers/generic/main.py
Comment (eskultety) [inline cachi2/core/package_managers/generic/main.py:64]: nitpick: I think this commentary is redundant as we pretty much use `join_within_root` for this particular purpose across the whole code base.

File: tests/unit/package_managers/test_generic.py
Comment (eskultety) [inline tests/unit/package_managers/test_generic.py:86]: I'd also add one with a complex download URL to test path extraction that is to serve as the target file name.

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py:92]: It would, but I know that these lockfiles won't be small and could go up to a couple thousand. I think that having a little more complicated code to achieve parallel download is worth it in this case.

Comment (eskultety) [reply cachi2/core/package_managers/generic/main.py:92]: We do parallel pretty much everywhere exactly due to reasons @kosciCZ mentions, so it's preferable and there's precedent -> marking as resolved.

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py:50]: > should be taken care of with an after model validator (as field validators must take attribute order validation into consideration and so are painful to deal with) OR maybe there's a callback for the Field descriptor that could compute the value but then again, I don't think there's a different primitive apart from model validators that would not need to care about other attributes not having been validated/set yet.

My first idea with this was to do this directly in the `Field` with a lambda, but that can't see other fields in the model.

Do I understand correctly that you're suggesting I use the `after` model validator to set the value of `target` based on the `download_url` ? I've seen that suggested, but it felt somewhat as misusing the validator to set something. Although I am happy to implement it, if that's the preferred way to do it.

File: cachi2/core/package_managers/generic/main.py
Comment (slimreaper35) [reply cachi2/core/package_managers/generic/main.py:92]: ACK

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Fixed.

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Try except removed.

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Removed.

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py:97]: Separated into new commit

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Removed

File: tests/unit/package_managers/test_generic.py
Comment (kosciCZ) [reply tests/unit/package_managers/test_generic.py]: Added to the `LOCKFILE_VALID` case

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: added in `no_artifact_conflicts`

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py]: removed. Validator now directly sets the correct value for the target, as it is required in multiple places as resolved, as well as for validation of the entire lockfile.

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py]: Fixed.

File: cachi2/core/package_managers/generic/models.py
Comment (brunoapimentel) [inline cachi2/core/package_managers/generic/models.py:53]: I think that tying the class instantation to the existence of specific data in context is not ideal, but the alternative would be to either do the `target` validation within the `_resolve_generic_lockfile` (while also downloading the files), or even make the class mutable if we happen to need this info further down the line. 

So I'm fine with the current approach.

**Review by brunoapimentel — COMMENTED** [review]
Overall, LGTM, just left some small remarks.

File: cachi2/core/package_managers/generic/models.py
Comment (eskultety) [reply cachi2/core/package_managers/generic/models.py]: > Do I understand correctly that you're suggesting I use the after model validator to set the value of target based on the download_url ? I've seen that suggested, but it felt somewhat as misusing the validator to set something

Yes. I hear you, the only other way would be to deal with this from the caller, i.e. when the lockfile model is instantiated from YAML. When I look at the official docs https://docs.pydantic.dev/latest/concepts/validators/#using-validation-context-with-basemodel-initialization they use some overly complex mechanism to change mangle the value that the model is supposed to be instantiated with which gives me the confidence that setting reasonable default values is within the context of those validators since IIUC no other mechanism exists for that purpose in pydantic.

File: cachi2/core/package_managers/generic/models.py
Comment (slimreaper35) [inline cachi2/core/package_managers/generic/models.py:78]: Do these need to be sets ?

Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py:78]: Why not?

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py:53]: Thanks. There was a couple of options, but based on previous review comments, this seems to be the preferred option. I'm aware of the downsides that come with it.

File: tests/unit/package_managers/test_generic.py
Comment (kosciCZ) [reply tests/unit/package_managers/test_generic.py:151]: Good point, will do.

File: tests/integration/test_generic.py
Comment (kosciCZ) [reply tests/integration/test_generic.py:9]: I agree that we can drop the lockfile format related tests, as they are well covered with unit tests. However I'd argue that we should keep those that verify lockfile targets in some way, as those are pretty important to get right. I know that we use temp paths in unit tests, so nothing is mocked there, but I'd still feel better if integration tests also verified this. @eskultety Your opinion on this?

File: tests/unit/package_managers/test_generic.py
Comment (kosciCZ) [reply tests/unit/package_managers/test_generic.py:165]: Added

File: cachi2/core/package_managers/generic/models.py
Comment (slimreaper35) [reply cachi2/core/package_managers/generic/models.py:78]: Because you are checking if something already exists in the set. That seems a bit off to me. Sets do not contain duplicate values. That's why I raised the question.

File: cachi2/core/package_managers/generic/models.py
Comment (eskultety) [reply cachi2/core/package_managers/generic/models.py:25]: nitpick: Do we want this to be a map? I mean most commonly (those I have seen) are always string lists. I know that it probably makes it just a bit simpler to get a checksum object out of it, but I think plain strings are more common out there.

File: tests/integration/test_generic.py
Comment (eskultety) [reply tests/integration/test_generic.py:9]: Fully agree with @brunoapimentel that all of those but one tests are isolated to lockfile validation which doesn't integrate with any external functionality and so unit tests are sufficient.

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py:25]: Pairs is used here as a general concept, not a typing notation. I will change it to reflect L30 better.

> nitpick: Do we want this to be a map? I mean most commonly (those I have seen) are always string lists. I know that it probably makes it just a bit simpler to get a checksum object out of it, but I think plain strings are more common out there.

That feels like a comment that should have gone to the previous MR where the lockfile format was defined. Or the one with the design document, where it was also shown. I'm not opposed to it, if that's what the maintainers prefer, but I'd like you to think about the toll work that this creates for me to constantly rewrite stuff that seemed to be agreed upon beforehand.

File: cachi2/core/package_managers/generic/models.py
Comment (eskultety) [reply cachi2/core/package_managers/generic/models.py:53]: The unfortunate reality here is that it's a nested model, so to IMO do it by the book would mean to effectively validate the lockfile twice - once as a dict from the YAML parser, loop over all artifacts and override any missing targets and second via models. The former however would be relying on the structure which the pydantic models validate so we can't use that approach sadly, though contexts aren't the cleanest way, I agree (but hey, they showcase a similar thing in the docs!)

> alternative would be to either do the target validation within the _resolve_generic_lockfile

Doable and also kinda clean to be honest :), I'll leave it to @kosciCZ to decide.

> or even make the class mutable if we happen to need this info further down the line.

Honestly I'd rather keep using contexts for validators like this than go with mutable models, to me, these are the sources of truth for the rest of the logic so changing them may introduce hard to catch issues later on.

@kosciCZ depending on which approach you stick to, can you please update the commentary to be more verbose on the context need?

File: cachi2/core/package_managers/generic/models.py
Comment (eskultety) [reply cachi2/core/package_managers/generic/models.py:78]: Because sets have O(1) complexity on lookup which is much faster than searching a list (O(N)) in the loop below.

File: cachi2/core/package_managers/generic/models.py
Comment (eskultety) [reply cachi2/core/package_managers/generic/models.py:80]: > This way all issues will be detected in one pass.

Not bad, this kind of handling is what I've missed in so many tools I have used in my life which only failed at the nearest problem.

File: cachi2/core/package_managers/generic/models.py
Comment (eskultety) [reply cachi2/core/package_managers/generic/models.py:25]: > I'm not opposed to it, if that's what the maintainers prefer, but I'd like you to think about the toll work that this creates for me to constantly rewrite stuff that seemed to be agreed upon beforehand.

That's fair, I was just curious, any of us can change it before we roll this one out as supported, it's a trivial change.

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py:55]: Update the comment to be more verbose

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py]: Improved description to better reflect typing information.

File: cachi2/core/package_managers/generic/models.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/models.py]: Very good suggestion. Slightly edited to not output parts of the error message when multiple conflicts are found. Thanks!

**Comment by kosciCZ** [comment]
The integration tests seem to be failing on something unrelated.

File: cachi2/core/package_managers/generic/models.py
Comment (a-ovchinnikov) [inline cachi2/core/package_managers/generic/models.py:85]:
```python
>>> "foo " * len([]) + "bar " * len([1])
'bar '
```
I had it in the original suggestion, but this works too (and is arguably more readable).

Comment (a-ovchinnikov) [reply cachi2/core/package_managers/generic/models.py:85]: Good catch! I was thinking about `* (len(urls) > 0)`, but did not type it out in fullness. `if..else` is a better way of expressing the idea anyway.

**Review by a-ovchinnikov — APPROVED** [review]
LGTM

**Review by slimreaper35 — APPROVED** [review]

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py:78]: Fixed.

**Comment by kosciCZ** [comment]
@brunoapimentel  @eskultety If you don't have any outstanding comments, please approve & merge. Thanks!

File: cachi2/core/package_managers/generic/main.py
Comment (eskultety) [inline cachi2/core/package_managers/generic/main.py]: You revert this hunk in the last commit. Sounds like a rebase typo.

File: cachi2/core/package_managers/generic/main.py
Comment (eskultety) [inline cachi2/core/package_managers/generic/main.py:75]: This should have happened in the same commit as the other error type fix.
To minimize the rebase conflicts and changes in general, please move the last commit to the very beginning of the series to take care of all these unrelated error changes and then introduce your actual topic changes.

**Review by eskultety — COMMENTED** [review]
Last minor comments, quick respin will do -> ACK.

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Yes, I messed that up during splitting of the commits, I meant to edit the below error. Let me fix that.

Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Fixed. The changes related to PackageManagerError -> PackageRejected are now self contained in 6520893cf5d2cab5650f227d84a91e6b0fc5b5f5 and is the first commit in this PR.

**Review by eskultety — APPROVED** [review]

**Comment by eskultety** [comment]
@brunoapimentel any further comments or can we merge?

**Review by brunoapimentel — APPROVED** [review]

**Pull request merged**
